fblock.go

```diff
@@ -0,0 +1,455 @@
+func (fb *FBlock) Get(c *Client) (err error) {
+	if fb.IsPopulated() {
+		return nil
+	}
+
+	if fb.KeyMR != nil {
+		orig := fb.KeyMR
+		defer func() { // Ensure we got the block we asked for
+			if err != nil {
+				return
+			}
+
+			if fb.KeyMR != nil && *orig != *fb.KeyMR {
+				err = fmt.Errorf("invalid key merkle root")
+				return
+			}
+		}()
+
+		params := struct {
+			Hash *Bytes32 `json:"hash"`
+		}{Hash: fb.KeyMR}
+		var result struct {
+			Data Bytes `json:"data"`
+		}
+		if err := c.FactomdRequest("raw-data", params, &result); err != nil {
+			return err
+		}
+		return fb.UnmarshalBinary(result.Data)
+	}
+
+	params := struct {
+		Height uint32 `json:"height"`
+	}{fb.Header.Height}
+	result := struct {
+		// We will ignore all the other fields, and just unmarshal from the raw.
+		RawData Bytes `json:"rawdata"`
+	}{}
+	if err := c.FactomdRequest("fblock-by-height", params, &result); err != nil {
+		return err
+	}
+
+	return fb.UnmarshalBinary(result.RawData)
+}
+
+const (
+	FBlockMinHeaderLen = 32 + // Factoid ChainID
+		32 + // BodyMR
+		32 + // PrevKeyMR
+		32 + // PrevLedgerKeyMR
+		8 + // EC Exchange Rate
+		4 + // DB Height
+		0 + // Header Expansion size (varint)
+		0 + // Header Expansion Area (Min 0)
+		4 + // Transaction Count
+		4 // Body Size
+
+	FBlockMinTotalLen = FBlockMinHeaderLen
+)
+
+// UnmarshalBinary unmarshals raw directory block data.
+//
+// Header
+// [Factoid Block ChainID (Bytes32{31:0x0a})] +
+// [BodyMR (Bytes32)] +
+// [PrevKeyMR (Bytes32)] +
+// [PrevLedgerKeyMR (Bytes32)] +
+// [Exchange Rate (8 bytes)] +
+// [DB Height (4 bytes)] +
+// [Header Expansion size (Bytes)] +
+// [Header Expansion Area (Bytes)] +
+// [Transaction Count (4 bytes)] +
+// [Body Size (4 bytes)] +
+//
+// Body
+// [Tx 0 (Bytes)] +
+// ... +
+// [Tx N (Bytes)] +
+//
+// https://github.com/FactomProject/FactomDocs/blob/master/factomDataStructureDetails.md#factoid-block
+func (fb *FBlock) UnmarshalBinary(data []byte) (err error) {
+	// Because the length of an FBlock is hard to define up front, we will catch
+	// any sort of out of bound errors in a recover
+	defer func() {
+		if r := recover(); r != nil {
+			err = fmt.Errorf("failed to unmarshal")
+		}
+	}()
+
+	// TODO: More length checks up front
+	if len(data) < FBlockMinTotalLen {
+		return fmt.Errorf("insufficient length")
+	}
+
+	if bytes.Compare(data[:32], factoidBlockChainID[:]) != 0 {
+		return fmt.Errorf("invalid factoid chainid")
+	}
+	i := 32
+	fb.Header.BodyMR = new(Bytes32)
+	i += copy(fb.Header.BodyMR[:], data[i:i+len(fb.Header.BodyMR)])
+	fb.Header.PrevKeyMR = new(Bytes32)
+	i += copy(fb.Header.PrevKeyMR[:], data[i:i+len(fb.Header.PrevKeyMR)])
+	fb.Header.PrevLedgerKeyMR = new(Bytes32)
+	i += copy(fb.Header.PrevLedgerKeyMR[:], data[i:i+len(fb.Header.PrevLedgerKeyMR)])
+	fb.Header.ExchangeRate = binary.BigEndian.Uint64(data[i : i+8])
+	i += 8
+	fb.Header.Height = binary.BigEndian.Uint32(data[i : i+4])
+	i += 4
+
+	expansionsize, read := binary.Varint(data[i:])
```

Comment:
That varint use was a mistake, good catch. I probably let my IDE choose that varint implementation. The idea behind the current unmarshal is that it won't break your code if you do not update.

In the admin block for example, we wrote a marshal function that handles all future admin block types. So when you decode an admin block, if you do not understand it, you just pack it into a byte slice and ignore it.

For this section, if someone if using your library and following fblocks, if the expansion stuff is used, it won't break any clients. They will still be able to read factoid txs, because the marshal will just pack that away, and ignore it.